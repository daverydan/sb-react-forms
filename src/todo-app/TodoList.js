import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";

const TodoList = () => {
  const INITIAL_STATE = ["clean room", "eat lunch", "take nap"];
  const [todos, setTodos] = useState(INITIAL_STATE);
  const addTodo = (todo) => {
    setTodos([...todos, todo]);
  };
  const removeTodo = (idx) => {
    const newTodos = [...todos]; // clone
    newTodos.splice(idx, 1); // remove
    setTodos(newTodos); // set
  };
  return (
    <>
      <NewTodoForm addTodo={addTodo} />
      <hr />
      <div className="TodoList" data-testid="todo-list">
        {todos.map((todo, idx) => (
          <Todo key={idx} id={idx} todo={todo} removeTodo={() => removeTodo(idx)} />
        ))}
      </div>
    </>
  );
};

export default TodoList;
