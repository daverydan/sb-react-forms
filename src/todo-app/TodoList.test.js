import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders w/out crashing", () => {
  render(<TodoList />);
});

test("it matches snapshot", () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

test("should add new item", () => {
  const { queryByText, queryByLabelText, getByTestId } = render(<TodoList />);
  const input = queryByLabelText("Add Task");
  const button = queryByText("Add");
  const todoList = getByTestId("todo-list");
  expect(queryByText("wake up")).not.toBeInTheDocument();
  fireEvent.change(input, { target: { value: "wake up" } });
  fireEvent.click(button);
  expect(queryByText("wake up")).toBeInTheDocument();
  expect(todoList.children.length).toBe(4);
});

test("should delete an item", () => {
  const { getByTestId, queryByText } = render(<TodoList />);
  const button = getByTestId("todo-btn-2");
  fireEvent.click(button);
  expect(queryByText("take nap")).not.toBeInTheDocument();
});
