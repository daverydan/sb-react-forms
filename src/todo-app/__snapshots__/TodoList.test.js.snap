// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewTodoForm"
  >
    <label
      for="todo"
    >
      Add Task
    </label>
    <input
      id="todo"
      name="todo"
      type="text"
      value=""
    />
    <button>
      Add
    </button>
  </form>
  <hr />
  <div
    class="TodoList"
    data-testid="todo-list"
  >
    <div
      class="Todo"
    >
      clean room 
      <button
        data-testid="todo-btn-0"
      >
        x
      </button>
    </div>
    <div
      class="Todo"
    >
      eat lunch 
      <button
        data-testid="todo-btn-1"
      >
        x
      </button>
    </div>
    <div
      class="Todo"
    >
      take nap 
      <button
        data-testid="todo-btn-2"
      >
        x
      </button>
    </div>
  </div>
</DocumentFragment>
`;
